### PR TITLE
Add mutation endpoint for requestType

### PR DESCRIPTION
**Pull request by kevin-wangg**
- Added createRequestType endpoint for creating requestTypes
  - To create a requestType, pass in the requestType object. Only the name field must be provided, other fields can resort to their defaults

- Added updateRequestType endpoint for updating requestTypes
  - To update a requestType, pass in the requestType object and the id of the requestType to be updated
    - Only the fields that need to be updated needed to be passed in the object
    - To update the requests field, all three arrays must be given (ie. you can't directly modify a subfield because I couldn't figure out how to do that)

- Added softDeleteRequestType endpoint for soft deleting requestTypes (change the deleted field to true)

Screenshot from 2021-03-14 23-09-04
   
Screenshot from 2021-03-14 23-08-29

Screenshot from 2021-03-14 20-22-28

### CODE REVIEW

**Review by Anand-Issac — COMMENTED** [review]
I noticed that creating a user provides the response message as "RequestType successfully updated" instead of "RequestType successfully created".

File: server/graphql/resolvers.ts [hunk 1]
```diff
@@ -15,6 +16,10 @@ const resolvers = {
     requestGroup: (_, { id }, { dataSources }): RequestGroupInterface => dataSources.requestGroups.getById(Types.ObjectId(id)),
     requestGroups: (_, __, { dataSources }): Array<RequestGroupInterface> => dataSources.requestGroups.getAll()
   },
+  Mutation: {
+    createOrUpdateRequestType: (_, { newRequestType, id }, { dataSources }): Promise<ServerResponseInterface> => dataSources.requestTypes.createOrUpdateRequestType(newRequestType, id),
+    softDeleteRequestType: (_, { id }, { dataSources}): Promise<ServerResponseInterface> => dataSources.requestTypes.softDeleteRequestType(id)
```
Comment (guy-ro) [inline server/graphql/resolvers.ts]: please replace endpoint with deleteRequestType since the user of the API doesn't care if it is delete/softDelete

you can more succinctly call createOrUpdateRequestType as upsertRequestType or modifyRequestType, but no worries

File: server/datasources/requestTypeDataSource.ts [hunk 2]
```diff
@@ -1,9 +1,80 @@
...
+import { Types } from 'mongoose'
 
 export default class RequestTypeDataSource extends CachedMongooseDataSource<RequestTypeDocument> {
   constructor() {
     super(RequestTypeCache)
   }
+  async softDeleteRequestType(id: Types.ObjectId) {
+    const RequestType = this.cache.model
+    try {
+      const promise = RequestType.findByIdAndUpdate(id, {"deleted": true})
```
Comment (guy-ro) [inline server/datasources/requestTypeDataSource.ts]: you can catch errors here with .then().catch(). Otherwise, the async function may return promise before the query results in an error

File: server/datasources/requestTypeDataSource.ts [hunk 3]
```diff
@@ -1,9 +1,80 @@
 import CachedMongooseDataSource from './cachedMongooseDataSource'
```
Comment (guy-ro) [inline server/datasources/requestTypeDataSource.ts]: if the functions here are generic to all data sources, then please put these functions in the cachedMongooseDataSource

File: server/datasources/requestTypeDataSource.ts [hunk 4]
```diff
@@ -1,9 +1,80 @@
...
+    catch(error) {
+      console.log(error)
+      return {
+        "success": false,
+        "message": error._message,
+        "id": id
+      }
+    }
+  }
+  async createOrUpdateRequestType(newRequestType: RequestTypeDocument, id: Types.ObjectId): Promise<ServerResponseInterface> {
```
Comment (guy-ro) [inline server/datasources/requestTypeDataSource.ts]: please change "newRequestType" to "requestType" to better reflect the purpose of the function

File: server/datasources/requestTypeDataSource.ts [hunk 5]
```diff
@@ -1,9 +1,80 @@
...
+    const RequestType = this.cache.model
+    if(id) {
```
Comment (guy-ro) [inline server/datasources/requestTypeDataSource.ts]: There are many approaches to this, but ideally, call findByIdAndUpdate with upsert and then based on response, react accordingly to whether it was insert or update.

See https://stackoverflow.com/questions/35059580/mongoose-how-to-check-if-document-is-modified-via-model-findoneandupdate for one way to do this.

That is, don't have a case if(id), and just upsert the passed in requestType

File: server/datasources/requestTypeDataSource.ts [hunk 6]
```diff
@@ -1,9 +1,80 @@
...
+      catch(error) {
+        console.log(error)
+        return {
+          "success": false,
+          "message": error._message,
+          "id": id
+        }
+      }
+    }
+  }
```
Comment (guy-ro) [inline server/datasources/requestTypeDataSource.ts]: On create/update, we should call a function that adds the requestType to the requestGroup in the requestType's requestGroup field, and similarly removes it from the old one (if it exists) - check the cached requestGroups prior to calling mongoose.

Feel free to implement this when the requestGroup mutations are started

File: server/graphql/schema.ts [hunk 7]
```diff
@@ -53,6 +58,20 @@ const typeDefs = gql`
...
+    type Mutation {
+        createOrUpdateRequestType(newRequestType: RequestTypeInput, id: ID): ServerResponse
+        softDeleteRequestType(id: ID): ServerResponse
+    }
+    input RequestListingInput {
+        open: [ID]
+        fulfilled: [ID]
+        deleted: [ID]
+    }
+    input RequestTypeInput {
```
Comment (guy-ro) [inline server/graphql/schema.ts:74]: for the dateUpdated field, simply set it to Date.now when you upsert

File: server/models/serverResponseModel.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import { Types } from 'mongoose'
```
Comment (guy-ro) [inline server/models/serverResponseModel.ts]: this is not a mongoose model, so place it somewhere related to graphql

**Review by guy-ro — CHANGES_REQUESTED** [review]
See comments for changes to make

File: server/graphql/resolvers.ts (hunk 1 above)
Comment (kevin-wangg) [reply server/graphql/resolvers.ts]: I wanted to call it softDelete so it's clear from the developer's perspective that it doesn't actually delete it. Also so if we need to implement a hard delete in the future, it's clear there's a difference.
I'm open to changing it too though.

Comment (guy-ro) [reply server/graphql/resolvers.ts]: that's a good point

maybe softDelete will be more descriptive for future developers on this project

File: server/datasources/requestTypeDataSource.ts (hunk 5 above)
Comment (kevin-wangg) [reply server/datasources/requestTypeDataSource.ts]: I tried doing it this way but I ran into a couple of issues. 

Firstly, if id isn't provided (which should be the case when a user wants to create a new requestType), I'm not exactly sure how mongoose handles it. From my testing, it looked like nothing happened, as in no new requestType was created.

Secondly, even assuming the id being null case worked as we wanted, the two scenarios for creating and updating would need to be handled differently. For creating a new requestType, we have the option of not passing certain fields (and having them resort to their defaults) and this requires us to create a RequestType model out of the object passed in. But if we do the same thing when we update an existing requestType object, creating a RequestType model out of the object passed in will generate a new ObjectId, then mongoose throws an error cuz you're rewriting an immutable field. We could pass in the existing id when we create the RequestType object but then in the case where id is null, we would be creating an requestType with a null ObjectId. Alternatively, we could not create the RequestType model at all but then we would require every field be passed in when creating a new object which I felt made it not as useful. 

Basically combining the two together was not worth the work it took and I thought it'd be simpler to just separate the two cases. 

Let mw know if there is something I'm missing or if this didn't make sense lol

File: server/models/serverResponseModel.ts (hunk 8 above)
Comment (kevin-wangg) [reply server/models/serverResponseModel.ts]: Moved.

File: server/datasources/requestTypeDataSource.ts (hunk 5 above)
Comment (guy-ro) [reply server/datasources/requestTypeDataSource.ts]: In that case, it makes sense to split everything into two fn's, one for create and one for update, and similarly two resolvers. Note that for create, certain fields are required (see schema).

Comment (kevin-wangg) [reply server/datasources/requestTypeDataSource.ts]: Sounds good. Just want to @meganniu so she's aware of this change and if she has any input.

File: server/graphql/schema.ts (hunk 7 above)
Comment (kevin-wangg) [reply server/graphql/schema.ts:74]: Updated.

File: server/datasources/requestTypeDataSource.ts (hunk 3 above)
Comment (kevin-wangg) [reply server/datasources/requestTypeDataSource.ts]: Moved the softDelete function.

Comment (Anand-Issac) [reply server/datasources/requestTypeDataSource.ts]: I personally think it's not a good design choice to move softDeleteRequestType to cachedMongooseDataSource because it cannot be generically used by all data sources.

File: server/graphql/schema.ts [hunk 9]
```diff
@@ -53,6 +58,21 @@ const typeDefs = gql`
         requestGroup(id: ID): RequestGroup
         requestGroups: [RequestGroup]
     }
+    type Mutation {
+        createRequestType(requestType: RequestTypeInput): ServerResponse
```
Comment (Anand-Issac) [inline server/graphql/schema.ts:62]: In the schema the type of requestType is RequestTypeInput, but in the datasource the type is RequestTypeDocument (which basically is the RequestTypeInterface). I was wondering if this would cause any problems.

**Review by Anand-Issac — CHANGES_REQUESTED** [review]

File: server/datasources/requestTypeDataSource.ts (hunk 3 above)
Comment (guy-ro) [reply server/datasources/requestTypeDataSource.ts]: Every cachedMongooseDataSource uses soft deletion except for clients (and they probably should). @meganniu, should we add a deleted field to client objects? 

If so, then the behaviour is the same for all data sources. If not, then we should at least write a typescript generic function for softDeletion since behaviour is the same for three data sources (requests, requestTypes, requestGroups).

File: server/datasources/requestTypeDataSource.ts (hunk 2 above)
Comment (kevin-wangg) [reply server/datasources/requestTypeDataSource.ts]: Updated.

File: server/datasources/requestTypeDataSource.ts (hunk 4 above)
Comment (kevin-wangg) [reply server/datasources/requestTypeDataSource.ts]: Updated.

File: server/datasources/cachedMongooseDataSource.ts [hunk 10]
```diff
@@ -10,6 +10,27 @@ export default class CachedMongooseDataSource<DocumentType extends Document> ext
       this.cache = cache
   }
 
+  async softDeleteRequestType(id: Types.ObjectId) {
```
Comment (meganniu) [inline server/datasources/cachedMongooseDataSource.ts]: Let's make the function name and variable names generic. 
Replace `softDeleteRequestType` with `softDeleteById` (`resolvers.ts` will need to be changed too)
Directly use `this.cache.model` instead of `RequestType` on line 15

```
async softDeleteById(id: Types.ObjectId) {
  const promise = this.cache.model.findByIdAndUpdate(id, {"deleted": true})
    .then(res => {
      return {
        "success": true,
        "message": `${this.cache.name} successfully deleted`,
        "id": id
      }
    })
    .catch(error => {
      console.log(error)
      return {
        "success": false,
        "message": error._message,
        "id": id
      }
    })
  return promise
}
```

File: server/datasources/requestTypeDataSource.ts (hunk 3 above)
Comment (meganniu) [reply server/datasources/requestTypeDataSource.ts]: Yes let's add delete to client. we should be able to perform CRUD on all objects
@kevinwang2040

File: server/datasources/requestTypeDataSource.ts (hunk 6 above)
Comment (meganniu) [reply server/datasources/requestTypeDataSource.ts]: +1 this can def be another PR if you want

**Review by meganniu — APPROVED** [review]
@kevinkwchan wowow this is great tysm!
Just a few comments, but otherwise amazing work!!!

@guy-ro @Anand-Issac ty guys as well for helping out 💗

File: server/datasources/requestTypeDataSource.ts (hunk 3 above)
Comment (meganniu) [reply server/datasources/requestTypeDataSource.ts]: Oh also, i don't think we should make `create` and `update` functions generic
We need to perform type-specific validation and manipulation before creating/updating (e.g. check RequestGroup with id = `requestType.requestGroup` exists, remove RequestType id from old ReqestGroup when updating)

Comment (meganniu) [reply server/datasources/requestTypeDataSource.ts]: Actually maybe we could make `create` and `update` function generic.
We pass a function that performs type-specific validation and manipulation into `create` and `update`
```
create(object, validate: (object) => boolean) {
  if(validate(object)){
    create object
  }
}
```

@guy-ro @kevinkwchan @Anand-Issac thoughts?

Comment (guy-ro) [reply server/datasources/requestTypeDataSource.ts]: We can also put the logic in the resolver instead of in datasources. That way datasources are only involved in making queries to DB, not making multiple queries to multiple collections.

Comment (meganniu) [reply server/datasources/requestTypeDataSource.ts]: ^ i like the idea of putting the logic in resolvers better. 
@kevinkwchan @Anand-Issac how does that sound for you guys?

File: server/datasources/requestTypeDataSource.ts (hunk 5 above)
Comment (meganniu) [reply server/datasources/requestTypeDataSource.ts]: wow i completely missed this convo (how??)

The reasoning for splitting up the functions is pretty solid. sounds good to me!

@kevinwang2040 thnx for investigating mongoose's behaviour so thoroughly 🕵️
@guy-ro thnx (as always) for sharing ur ideas 🧠

File: server/datasources/requestTypeDataSource.ts (hunk 3 above)
Comment (Anand-Issac) [reply server/datasources/requestTypeDataSource.ts]: @meganniu  that sounds good to me

Comment (Anand-Issac) [reply server/datasources/requestTypeDataSource.ts]: > I personally think it's not a good design choice to move softDeleteRequestType to cachedMongooseDataSource because it cannot be generically used by all data sources.

jk

**Review by Anand-Issac — APPROVED** [review]
Everything lgtm!!

File: server/datasources/requestTypeDataSource.ts [hunk 11]
```diff
@@ -1,9 +1,44 @@
...
+import { ServerResponseInterface } from '../graphql/serverResponse'
+
+import { Types } from 'mongoose'
 
 export default class RequestTypeDataSource extends CachedMongooseDataSource<RequestTypeDocument> {
   constructor() {
     super(RequestTypeCache)
   }
+  
+  async createRequestType(requestType: RequestTypeDocument): Promise<ServerResponseInterface> {
```
Comment (guy-ro) [inline server/datasources/requestTypeDataSource.ts]: can't the user of this data source just call super.create directly?

Comment (guy-ro) [reply server/datasources/requestTypeDataSource.ts]: similar for the other functions

Comment (kevin-wangg) [reply server/datasources/requestTypeDataSource.ts]: hmm that's true. althought we don't have an instance of the cachedMongooseDataSource so i'm not sure how we would do that.

**Review by guy-ro — APPROVED** [review]
